### PR TITLE
Task5, DONE

### PR DESCRIPTION
**Pull request by multeng**
1. Task - [description](https://github.com/EPAM-JS-Competency-center/cloud-development-course-initial/blob/main/task5-import-to-s3/task.md)
2. Crosscheck - [information](https://github.com/EPAM-JS-Competency-center/cloud-development-course-initial/blob/main/task5-import-to-s3/task.md)
3. Deploy - [CloudFront link](http://dyhwz91va2tyf.cloudfront.net/), [Import](https://1mf0s2hzm0.execute-api.eu-west-1.amazonaws.com/dev/import?name=book1.csv)
4. Done - 22.09.21, Deadline: 22.09.2021 23:59 GMT+3
5. 7 points / 8 points (without tests :( )

### CODE REVIEW

**Comment by GEnextM** [comment]
Могу помочь с корсом
GEnext(Никита)#7387 - Discord

**Comment by multeng** [comment]
> Могу помочь с корсом
> GEnext(Никита)#7387 - Discord

Спасибо!), вчера ночью разобрался, парсер перенес на сегодня, доделал, можно проверять)

File: upload-service/serverless.ts [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    },
+    environment: {
+      AWS_NODEJS_CONNECTION_REUSE_ENABLED: '1',
+    },
+    lambdaHashingVersion: '20201221',
+    iamRoleStatements: [
+      {
+        Effect: 'Allow',
+        Action: ['s3:ListBucket'],
+        Resource: 'arn:aws:s3:::thrift-shop-upload',
```
Comment (egterent) [inline upload-service/serverless.ts:33]: I think it's better to move this to .env

File: upload-service/src/libs/s3Helper.ts [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+import * as csvParser from 'csv-parser';
+
+config.update({ region: 'eu-west-1', signatureVersion: 'v4' });
+
+const s3 = new S3();
+
+export const getSignedUrl = async (filename: string) => {
+  try {
+    return await s3.getSignedUrlPromise('putObject', {
+      Bucket: 'thrift-shop-upload',
```
Comment (egterent) [inline upload-service/src/libs/s3Helper.ts:11]: Could you please make this an enviroment variable

File: upload-service/src/functions/importFileParser/handler.ts [hunk 3]
```diff
@@ -0,0 +1,17 @@
+import { middyfy } from '@libs/lambda';
+import { formatJSONResponse } from '@libs/apiGateway';
+import { parse } from '@libs/s3Helper';
```
Comment (egterent) [inline upload-service/src/functions/importFileParser/handler.ts:3]: It's the nice idea to move the parsing logic to the separate module

File: upload-service/src/libs/s3Helper.ts [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+  try {
+    s3.getObject({ Bucket: 'thrift-shop-upload', Key: key })
+      .createReadStream()
+      .pipe(csvParser())
+      .on('error', (error) => console.log('Error: ', error))
+      .on('data', (row) => console.log('Row: ', row))
+      .on('end', () => console.log('Parse done'));
+
+    copyObject(key);
+    deleteObject(key);
```
Comment (egterent) [inline upload-service/src/libs/s3Helper.ts:60]: You have to use await here as copyObject and deleteObject are both async functions

File: upload-service/src/functions/importFileParser/handler.ts [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+
+const importFileParser = async (event) => {
+  try {
+    const results = [];
+    for (const record of event.Records) {
+      results.push(parse(record.s3.object.key));
```
Comment (egterent) [inline upload-service/src/functions/importFileParser/handler.ts:9]: The await statement is needed because parse is an async function

Comment (multeng) [reply upload-service/src/functions/importFileParser/handler.ts]: Fixed.

File: upload-service/src/libs/s3Helper.ts (hunk 4 above)
Comment (multeng) [reply upload-service/src/libs/s3Helper.ts]: Fixed.

File: upload-service/src/libs/s3Helper.ts (hunk 2 above)
Comment (multeng) [reply upload-service/src/libs/s3Helper.ts]: Fixed.

File: upload-service/serverless.ts (hunk 1 above)
Comment (multeng) [reply upload-service/serverless.ts]: Fixed.

File: upload-service/src/functions/importFileParser/handler.ts (hunk 3 above)
Comment (multeng) [reply upload-service/src/functions/importFileParser/handler.ts:3]: Should i move parse function from s3Helper.ts into parse.ts(create it)?

Comment (egterent) [reply upload-service/src/functions/importFileParser/handler.ts:3]: Sorry for a confusing comment. I just wanted to say that the way you implemented things here is just grate

**Pull request merged**